includes/sanitizers/class-amp-script-sanitizer.php

```diff
@@ -17,24 +21,167 @@
+	 *
+	 * @var array {
+	 *      @type bool $sanitize_scripts Whether to sanitize JS scripts (and not defer for final sanitizer).
+	 *      @type bool $unwrap_noscripts Whether to unwrap noscript elements.
+	 * }
+	 */
+	protected $args;
+
+	/**
+	 * Number of scripts which were kept.
 	 *
-	 * @todo Eventually this try to automatically convert script tags to AMP when they are recognized. See <https://github.com/ampproject/amp-wp/issues/1032>.
-	 * @todo When a script has an adjacent noscript, consider removing the script here to prevent validation error later. See <https://github.com/ampproject/amp-wp/issues/1213>.
+	 * This is used to determine whether noscript elements should be unwrapped.
+	 *
+	 * @var int
+	 */
+	protected $kept_script_count = 0;
+
+	/**
+	 * Style sanitizer.
+	 *
+	 * @var AMP_Style_Sanitizer
+	 */
+	protected $style_sanitizer;
+
+	/**
+	 * Image sanitizer.
+	 *
+	 * @var AMP_Img_Sanitizer
+	 */
+	protected $img_sanitizer;
+
+	/**
+	 * Form sanitizer.
+	 *
+	 * @var AMP_Form_Sanitizer
+	 */
+	protected $form_sanitizer;
+
+	/**
+	 * Init.
+	 *
+	 * @param AMP_Base_Sanitizer[] $sanitizers Sanitizers.
+	 */
+	public function init( $sanitizers ) {
+		parent::init( $sanitizers );
+
+		if (
+			array_key_exists( AMP_Style_Sanitizer::class, $sanitizers )
+			&&
+			$sanitizers[ AMP_Style_Sanitizer::class ] instanceof AMP_Style_Sanitizer
+		) {
+			$this->style_sanitizer = $sanitizers[ AMP_Style_Sanitizer::class ];
+		}
+
+		if (
+			array_key_exists( AMP_Img_Sanitizer::class, $sanitizers )
+			&&
+			$sanitizers[ AMP_Img_Sanitizer::class ] instanceof AMP_Img_Sanitizer
+		) {
+			$this->img_sanitizer = $sanitizers[ AMP_Img_Sanitizer::class ];
+		}
+
+		if (
+			array_key_exists( AMP_Form_Sanitizer::class, $sanitizers )
+			&&
+			$sanitizers[ AMP_Form_Sanitizer::class ] instanceof AMP_Form_Sanitizer
+		) {
+			$this->form_sanitizer = $sanitizers[ AMP_Form_Sanitizer::class ];
+		}
+	}
+
+	/**
+	 * Sanitize script and noscript elements.
 	 *
 	 * @since 1.0
 	 */
 	public function sanitize() {
-		$noscripts = $this->dom->getElementsByTagName( 'noscript' );
+		if ( ! empty( $this->args['sanitize_scripts'] ) ) {
+			$this->sanitize_script_elements();
+		}
+
+		// If custom scripts were kept (after sanitize_js_script_elements ran) it's important that noscripts not be
+		// unwrapped or else this could result in the JS and no-JS fallback experiences both being present on the page.
+		// So unwrapping is only done no custom scripts were retained (and the sanitizer arg opts-in to unwrap).
+		if ( 0 === $this->kept_script_count && ! empty( $this->args['unwrap_noscripts'] ) ) {
+			$this->unwrap_noscript_elements();
+		}
+
+		// When there are kept custom scripts, skip tree shaking since it's likely JS will toggle classes that have
+		// associated style rules.
+		// @todo There should be an attribute on script tags that opt-in to keeping tree shaking and/or to indicate what class names need to be included.
```

Comment:
Wouldn't it be better to keep tree shaking enabled by default, and then provide an opt-in attribute to disable tree shaking?

Reply:
My thinking here is this, consider this common script in themes:

```html
<script>document.documentElement.className = document.documentElement.className.replace( 'no-js', 'js' );</script>
```

If there are CSS rules such as:

```
.js .mobile-nav > ul {
    display: none;
}
.no-js .mobile-nav > ul {
    display: block;
}
```

The result is that with tree-shaking, the second rule will get stripped out because it is not aware of the `className` change by the script. So this is why tree shaking needs to be disabled by default. Nevertheless, there could be a way to opt-in to keeping tree shaking by adding an attribute to the `script` to indicate which class names it would be mutating:

```html
<script data-mutation-classlist="no-js js">
document.documentElement.className = document.documentElement.className.replace( 'no-js', 'js' );
</script>
```

If such a hypothetical attribute were present, then those class names could be added to the `AMP_Style_Sanitizer::$used_class_names` and the style rule would not get stripped out by the tree shaker.

So that's the thinking behind this comment.

Reply:
Ah OK that makes sense 👍.